File: 백준/Silver/1058. 친구/친구.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String ipt;
        String[] ipts;

        int N = Integer.parseInt(br.readLine());

        boolean[][] connected = new boolean[N][N];
        boolean[][] graph = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            ipt = br.readLine();
            for (int j = 0; j < ipt.length(); j++) {
                if(ipt.charAt(j) == 'Y'){
                    graph[i][j] = true;
                    connected[i][j] = true;
                }
            }
        }

        int answer = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                for (int k = 0; k < N; k++) {
                    if(i == j) continue;
                    if(graph[i][j] || (graph[i][k] && graph[k][j])) connected[i][j] = true;
                }
            }
        }


        for (int i = 0; i < N; i++) {
            int cnt = 0;
            for (int j = 0; j < N; j++) {
                if(connected[i][j]) cnt++;
            }
            answer = Math.max(cnt, answer);
        }

        System.out.println(answer);
    }
}

File: 백준/Silver/1058. 친구/친구.py
import sys

N = int(sys.stdin.readline())
graph = [list(sys.stdin.readline().strip()) for _ in range(N)]
is_connected = [[0 for __ in range(N)] for _ in range(N)]

for i in range(N):
    for j in range(N):
        for k in range(N):
            if i == j: continue
            if graph[i][j] == 'Y' or (graph[i][k] == 'Y' and graph[k][j] == 'Y'):
                is_connected[i][j] = 1

answer = -1
for row in is_connected:
    answer = max(answer, sum(row))

print(answer)